server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -345,7 +383,30 @@ public static boolean tryFastFilterAggregation(
         final BiConsumer<Long, Integer> incrementDocCount
     ) throws IOException {
         if (fastFilterContext == null) return false;
-        if (fastFilterContext.filters == null) return false;
+        if (!fastFilterContext.rewriteable) return false;
+
+        NumericDocValues docCountValues = DocValues.getNumeric(ctx.reader(), DocCountFieldMapper.NAME);
+        if (docCountValues.nextDoc() != NO_MORE_DOCS) {
+            logger.debug(
+                "Shard {} segment {} has at least one document with _doc_count field, skip fast filter optimization",
+                fastFilterContext.context.indexShard().shardId(),
+                ctx.ord
+            );
+            return false;
+        }
+
+        // if no filters built at shard level (see getDateHistoAggBounds method for possible reasons)
+        // check if the query is functionally match-all at segment level
+        if (!fastFilterContext.filtersBuiltAtShardLevel && !segmentMatchAll(fastFilterContext.context, ctx)) return false;
+        if (!fastFilterContext.filterBuilt()) {
+            logger.debug(
+                "Shard {} segment {} functionally match all documents. Build the fast filter",
+                fastFilterContext.context.indexShard().shardId(),
+                ctx.ord
+            );
+            fastFilterContext.buildFastFilter(FastFilterRewriteHelper::getIndexBounds);
+        }
```

Comment:
Logic at line 124 for check functionally match all is pushed down here @msfroh. Please take a look.